src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,352 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-data-model-provider/EmberDataBuffer.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage-null-handling.h>
+#include <app/util/odd-sized-integers.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLVTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <protocols/interaction_model/Constants.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <limits>
+
+namespace chip {
+namespace app {
+namespace Ember {
+
+namespace {
+
+/// Maximum length of a string, inclusive
+///
+/// the max size value (0xFF and 0xFFFF) is reserved for NULL representation so
+/// it is not available
+constexpr uint32_t MaxLength(EmberAttributeBuffer::PascalStringType s)
+{
+    if (s == EmberAttributeBuffer::PascalStringType::kShort)
+    {
+        return std::numeric_limits<uint8_t>::max() - 1;
+    }
+    // EmberAttributeBuffer::PascalStringType::kLong:
+    return std::numeric_limits<uint16_t>::max() - 1;
+}
+
+struct UnsignedDecodeInfo
+{
+    unsigned byteCount;
+    uint64_t maxValue;
+};
+
+constexpr UnsignedDecodeInfo GetUnsignedDecodeInfo(EmberAfAttributeType type)
+{
+
+    switch (type)
+    {
+    case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return UnsignedDecodeInfo{ 1, NumericLimits::MaxUnsignedValue(1) };
+    case ZCL_INT16U_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
+        return UnsignedDecodeInfo{ 2, NumericLimits::MaxUnsignedValue(2) };
+    case ZCL_INT24U_ATTRIBUTE_TYPE: // Unsigned 24-bit integer
+        return UnsignedDecodeInfo{ 3, NumericLimits::MaxUnsignedValue(3) };
+    case ZCL_INT32U_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
+        return UnsignedDecodeInfo{ 4, NumericLimits::MaxUnsignedValue(4) };
+    case ZCL_INT40U_ATTRIBUTE_TYPE: // Unsigned 40-bit integer
+        return UnsignedDecodeInfo{ 5, NumericLimits::MaxUnsignedValue(5) };
+    case ZCL_INT48U_ATTRIBUTE_TYPE: // Unsigned 48-bit integer
+        return UnsignedDecodeInfo{ 6, NumericLimits::MaxUnsignedValue(6) };
+    case ZCL_INT56U_ATTRIBUTE_TYPE: // Unsigned 56-bit integer
+        return UnsignedDecodeInfo{ 7, NumericLimits::MaxUnsignedValue(7) };
+    case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return UnsignedDecodeInfo{ 8, NumericLimits::MaxUnsignedValue(8) };
+    }
+    chipDie();
+}
+
+struct SignedDecodeInfo
+{
+    unsigned byteCount;
+    int64_t minValue;
+    int64_t maxValue;
+};
+
+constexpr SignedDecodeInfo GetSignedDecodeInfo(EmberAfAttributeType type)
+{
+
+    switch (type)
+    {
+    case ZCL_INT8S_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return SignedDecodeInfo{
+            1,
+            NumericLimits::MinSignedValue(1),
+            NumericLimits::MaxSignedValue(1),
```

Comment:
Again, I think this would be more readable as `SignedDecodeInfo(1)` with the constructor handling the three places that byte count gets used....